File: app/weather_station/weather_station.py
import json
import random
import time
import sys
import os
from confluent_kafka import Producer


def delivery_report(err, msg):
    """Called once for each message produced to indicate delivery result."""
    if err is not None:
        print(f"Message delivery failed: {err}")
    else:
        print(f"Message delivered to {msg.topic()} [{msg.partition()}] at offset {msg.offset()}")


def create_weather_message(station_id, seq_number):
    """Create a weather status message with random values."""
    # Battery status with probabilities: low=30%, medium=40%, high=30%
    battery_status_options = ["low", "medium", "high"]
    battery_status_weights = [0.3, 0.4, 0.3]
    battery_status = random.choices(battery_status_options, battery_status_weights)[0]

    # Weather data with random values
    humidity = random.randint(10, 100)  # percentage
    temperature = random.randint(32, 110)  # Fahrenheit
    wind_speed = random.randint(0, 60)  # km/h

    # Create message
    message = {
        "station_id": station_id,
        "s_no": seq_number,
        "battery_status": battery_status,
        "status_timestamp": int(time.time()),
        "weather": {
            "humidity": humidity,
            "temperature": temperature,
            "wind_speed": wind_speed
        }
    }

    return message


def run_weather_station(station_id, kafka_bootstrap_servers="kafka:9092"):
    """Run the weather station simulation."""
    # Configure the Kafka producer
    conf = {
        'bootstrap.servers': kafka_bootstrap_servers,
        'client.id': f'weather-station-{station_id}'
    }
    producer = Producer(conf)

    # Topic to send messages to
    topic = "weather-data"

    # Sequence number counter
    seq_number = 1

    print(f"Weather Station {station_id} started. Sending data to Kafka topic: {topic}")

    try:
        while True:
            # Randomly drop messages at 10% rate
            if random.random() < 0.1:
                print(f"Station {station_id} - Dropped message {seq_number}")
                continue

            # Create weather message
            message = create_weather_message(station_id, seq_number)

            # Convert message to JSON string
            message_json = json.dumps(message)

            # Send message to Kafka
            producer.produce(topic=topic, key=str(station_id), value=message_json.encode('utf-8'), callback=delivery_report)

            # Force sending of all messages
            producer.flush()

            print(f"Station {station_id} - Sent message {seq_number}: {message_json}")

            # Increment sequence number
            seq_number += 1

            # Wait for 1 second before next reading
            time.sleep(1)
    except KeyboardInterrupt:
        print(f"Weather Station {station_id} stopping...")
    finally:
        # Clean up resources
        producer.flush()
        print(f"Weather Station {station_id} stopped")


def main():
    # Get station ID from command line or environment variable, default is 1
    station_id = int(sys.argv[1]) if len(sys.argv) > 1 else int(os.environ.get("STATION_ID", 1))

    # Get Kafka bootstrap servers from environment variable or use default
    kafka_bootstrap_servers = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "kafka:9092")
    run_weather_station(station_id, kafka_bootstrap_servers)


if __name__ == "__main__":
    main()

File: Weather-Station/src/main/java/com/geekcap/javaworld/WeatherStation.java
package com.geekcap.javaworld;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.Random;
import java.util.concurrent.ExecutionException;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import com.fasterxml.jackson.databind.ObjectMapper;

public class WeatherStation {
    private final long stationId;
    private long sequenceNumber;
    private final String kafkaBootstrapServers;
    private final Random random;
    private final ObjectMapper objectMapper;

    public WeatherStation(long stationId, String kafkaBootstrapServers) {
        this.stationId = stationId;
        this.sequenceNumber = 1;
        this.kafkaBootstrapServers = kafkaBootstrapServers;
        this.random = new Random();
        this.objectMapper = new ObjectMapper();
    }

    public void run() {
        // Configure the Kafka producer
        Properties properties = new Properties();
        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBootstrapServers);
        properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        properties.setProperty(ProducerConfig.CLIENT_ID_CONFIG, "weather-station-" + stationId);

        String topic = "weather-data";
        System.out.println("Weather Station " + stationId + " started. Sending data to Kafka topic: " + topic);

        try (KafkaProducer<String, String> producer = new KafkaProducer<>(properties)) {
            while (true) {
                try {
                    // Randomly drop messages at 10% rate
                    if (random.nextDouble() < 0.1) {
                        System.out.println("Station " + stationId + " - Dropped message " + sequenceNumber);
                        sequenceNumber++;
                        continue;
                    }

                    // Create weather message
                    WeatherMessage message = createWeatherMessage();

                    // Convert message to JSON string
                    String messageJson = objectMapper.writeValueAsString(message);

                    // Send message to Kafka
                    ProducerRecord<String, String> record = new ProducerRecord<>(
                            topic,
                            String.valueOf(stationId),
                            messageJson
                    );

                    producer.send(record, (metadata, e) -> {
                        if (e != null)
                            System.out.println("Message delivery failed: " + e.getMessage());
                        else
                            System.out.println("Message delivered to " + metadata.topic() +
                                    " [" + metadata.partition() + "] at offset " + metadata.offset());
                    }).get();

                    System.out.println("Station " + stationId + " - Sent message " + sequenceNumber + ": " + messageJson);

                    // Increment sequence number
                    sequenceNumber++;

                    // Wait for 1 second before next reading
                    Thread.sleep(1000);

                } catch (InterruptedException e) {
                    System.out.println("Weather Station " + stationId + " interrupted");
                    break;
                } catch (ExecutionException e) {
                    System.out.println("Error sending message: " + e.getMessage());
                } catch (Exception e) {
                    System.out.println("Unexpected error: " + e.getMessage());
                }
            }
        } catch (Exception e) {
            System.out.println("Error initializing producer: " + e.getMessage());
        } finally {
            System.out.println("Weather Station " + stationId + " stopped");
        }
    }

    private WeatherMessage createWeatherMessage() {
        // Battery status with probabilities: low=30%, medium=40%, high=30%
        String batteryStatus;
        double batteryRandom = random.nextDouble();
        if (batteryRandom < 0.3)
            batteryStatus = "low";
        else if (batteryRandom < 0.7)
            batteryStatus = "medium";
        else
            batteryStatus = "high";

        // Weather data with random values
        int humidity = random.nextInt(91) + 10;     // 10-100 percentage
        int temperature = random.nextInt(79) + 32;  // 32-110 Fahrenheit
        int windSpeed = random.nextInt(61);         // 0-60 km/h

        // Create message
        WeatherMessage message = new WeatherMessage();
        Map<String, Integer> weather = new HashMap<>();
        weather.put("humidity", humidity);
        weather.put("temperature", temperature);
        weather.put("wind_speed", windSpeed);

        message.stationId = this.stationId;
        message.sequenceNumber = this.sequenceNumber;
        message.batteryStatus = batteryStatus;
        message.statusTimestamp = ZonedDateTime.now(ZoneId.systemDefault()).toEpochSecond() * 1000L;
        message.weather = weather;

        return message;
    }

    static class WeatherMessage {
        public long stationId;
        public long sequenceNumber;
        public String batteryStatus;
        public long statusTimestamp;
        public Map<String, Integer> weather;

        // Default constructor required for Jackson
        public WeatherMessage() {}
    }
}
